### PR TITLE
Phone number validation

### PR DESCRIPTION
**Pull request by echen1031**
What was done:
1) Phone number validation 
2) Two-factor authentication using sms

### CODE REVIEW

**Comment by echen1031** [comment]
@alex-nexus  Hey alex, can you take

File: app/controllers/subscriptions_controller.rb [hunk 1]
```diff
@@ -58,7 +58,7 @@ def pause_subscription
   def start_verification
     result = Nexmo::Client.new.send_verification_request(number: @subscription.phone_number, brand: "OpenManna")
```
Comment (echen1031) [inline app/controllers/subscriptions_controller.rb]: @alex-nexus.  I don't know what's the proper way to test this external service.  Can we pair?

File: app/controllers/subscriptions_controller.rb [hunk 2]
```diff
@@ -44,13 +44,35 @@ def destroy
...
+  end
+
+  def pause
     @subscription.toggle!(:active)
     redirect_to subscriptions_path
   end
 
   private
 
+  def start_verification
```
Comment (alex-nexus) [inline app/controllers/subscriptions_controller.rb]: These 3 functions are written well but I think they can be moved to another service class, such as subscription_verifier

File: app/controllers/subscriptions_controller.rb [hunk 3]
```diff
@@ -44,13 +44,35 @@ def destroy
...
+    result = Nexmo::Client.new.send_verification_request(number: @subscription.phone_number, brand: "OpenManna")
+    if result['status'] == '0'
+      redirect_to edit_verification_path(id: result['request_id'], sub_id: @subscription.id)
+    else
+      redirect_to subscriptions_path
+      flash[:error] = 'Could not verify your number. Please contact support.'
+    end
+  end
+
+  def verify_phone_number
```
Comment (alex-nexus) [inline app/controllers/subscriptions_controller.rb]: IMO, 1 line function is too much, maybe combine these 2 one line functions

File: spec/features/user_manages_subscription_spec.rb [hunk 4]
```diff
@@ -70,9 +70,19 @@
   end
 
   scenario "pauses his subscription successfully" do
```
Comment (alex-nexus) [inline spec/features/user_manages_subscription_spec.rb:104]: Good job~

File: app/controllers/subscriptions_controller.rb [hunk 5]
```diff
@@ -11,7 +11,10 @@ def new
 
   def create
     @subscription = current_user.subscriptions.build(subscription_params)
-    if @subscription.save
+    if @subscription.over_limit?
+      flash[:error] = "Sorry, only two subscriptions per user are allowed at this time."
+      redirect_to subscriptions_path
+    elsif @subscription.save
```
Comment (alex-nexus) [inline app/controllers/subscriptions_controller.rb]: nice!

File: app/controllers/subscriptions_controller.rb [hunk 6]
```diff
@@ -11,7 +11,10 @@ def new
 
   def create
     @subscription = current_user.subscriptions.build(subscription_params)
-    if @subscription.save
+    if @subscription.over_limit?
```
Comment (alex-nexus) [inline app/controllers/subscriptions_controller.rb]: The question over_limit? should be on the user level, not on subscription itself

File: app/models/subscription.rb [hunk 7]
```diff
@@ -16,4 +17,12 @@ def current_day_in_words
     current_day = Time.now.strftime("%A").downcase
     "send_#{current_day}"
   end
+
+  def over_limit?
+    Subscription.where(user_id: self.user_id).size >= 2
+  end
+
+  def needs_verification?
+    self.active == false
```
Comment (alex-nexus) [inline app/models/subscription.rb]: one line function may be too much. You can simply do 
self.active? or @subscription.active?

File: app/services/sms_client.rb [hunk 8]
```diff
@@ -10,4 +10,14 @@ def initialize
   def send_message(to:, text:)
     @client.send_message(from: @from, to: to, text: text)
   end
+
+  def send_verification(number)
+    @client.send_verification_request(number: number, brand: "OpenManna")
```
Comment (alex-nexus) [inline app/services/sms_client.rb:15]: again 1 line function may be over kill, but just a suggestion

File: app/services/verification_request_checker.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+class VerificationRequestChecker
```
Comment (alex-nexus) [inline app/services/verification_request_checker.rb:1]: I am not sure if you need a service class for checker, it doesn't seem there is a lot of logic and all the functions are just based on the results

File: app/services/verification_request_sender.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
+class VerificationRequestSender
```
Comment (alex-nexus) [inline app/services/verification_request_sender.rb:1]: perhaps combine VerificationRequestSender and VerificationRequestChecker?

**Pull request merged**